src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -9,22 +9,70 @@
 #include "ch4_impl.h"
 #include "ch4r_proc.h"
 #include "ch4r_recv.h"
+#include "mpidu_genq.h"
 
-static inline void MPIDIG_prepare_recv_req(int rank, int tag, MPIR_Context_id_t context_id,
-                                           void *buf, MPI_Aint count, MPI_Datatype datatype,
-                                           MPIR_Request * rreq)
+void MPIDIG_deferred_am_irecv_dequeue(MPIDIG_deferred_am_irecv_req_t * dreq);
+int MPIDIG_deferred_am_irecv_issue(MPIDIG_deferred_am_irecv_req_t * dreq, int *status);
+
+static inline int MPIDIG_deferred_am_irecv_enqueue(int rank, int tag, int context_id, void *buf,
+                                                   MPI_Aint count, MPI_Datatype datatype,
+                                                   MPIR_Request * rreq, MPIR_Request * unexp_req);
+
+static inline int MPIDIG_prepare_recv_req(int rank, int tag, MPIR_Context_id_t context_id,
+                                          void *buf, MPI_Aint count, MPI_Datatype datatype,
+                                          MPIR_Request * rreq)
 {
+    int mpi_errno = MPI_SUCCESS;
+    MPL_pointer_attr_t attr;
+
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_PREPARE_RECV_REQ);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_PREPARE_RECV_REQ);
 
+    MPIR_GPU_query_pointer_attr(buf, &attr);
+
     MPIDIG_REQUEST(rreq, rank) = rank;
     MPIDIG_REQUEST(rreq, tag) = tag;
     MPIDIG_REQUEST(rreq, context_id) = context_id;
-    MPIDIG_REQUEST(rreq, datatype) = datatype;
-    MPIDIG_REQUEST(rreq, buffer) = (char *) buf;
-    MPIDIG_REQUEST(rreq, count) = count;
+    if (attr.type == MPL_GPU_POINTER_DEV) {
+        /* Force staging in registered memory if recv buf is on GPU */
+        MPI_Aint data_sz = 0;
+        MPIR_Datatype *dt_ptr = NULL;
+
+        MPIDIG_REQUEST(rreq, buffer_type) = MPIDIG_BUFFER_TYPE__TEMP;
+        /* temporary unpack buffer stored in (buffer, datatype, count) for recv into temp buffer
+         * and will unpack to user buffer later */
+        MPIDI_Datatype_get_size_dt_ptr(count, datatype, data_sz, dt_ptr);
+        MPIDIG_REQUEST(rreq, datatype) = MPI_BYTE;
+        MPIDIG_REQUEST(rreq, count) = data_sz;
+        /* init (user_buffer, user_datatype, user_count) with user buffer info */
+        MPIDIG_REQUEST(rreq, user_datatype) = datatype;
+        MPIDIG_REQUEST(rreq, user_buffer) = buf;
+        MPIDIG_REQUEST(rreq, user_count) = count;
+
+        if (data_sz <= MPIR_CVAR_CH4_AM_PACK_BUFFER_SIZE) {
+            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_global.exp_pack_buf_pool,
+                                                           MPIDIG_REQUEST(rreq, buffer));
+            if (MPIDIG_REQUEST(rreq, buffer) == NULL) {
+                mpi_errno = MPI_ERR_OTHER;
+            }
```

Comment:
Why is this error code being overridden compared to what `MPIDU_genq_private_pool_alloc_cell` already set?  Shouldn't you just jump to fn_fail if this happens, instead?